File: proposed-projects/round-trip-experiment/reverse/20260909T182714Z_2643e4be/20260910T100600Z_9e4cc6a6/ported_implementation/gbnf/grammar_graph/grammar_graph_types.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Dict, List, Optional, Union

Range = List[int]

ValidInput = Union[str, int, List[int]]


class Rule:
    @property
    def type(self) -> str:
        """
        The class name, as reported by this rule's `__dict__`.
        """
        return type(self).__name__

    @property  # type: ignore[misc]
    def __dict__(self) -> Dict[str, Any]:  # type: ignore[override]
        return {"type": type(self).__name__}

    def __eq__(self, other: object) -> bool:
        return isinstance(other, type(self))

    def __hash__(self) -> int:
        return id(self)

    def __repr__(self) -> str:
        return f"{type(self).__name__}()"

    def __str__(self) -> str:
        return f"{type(self).__name__}()"


class RuleWithValue(Rule):
    def __init__(self, value: Any) -> None:
        self.value = value

    @property  # type: ignore[misc]
    def __dict__(self) -> Dict[str, Any]:  # type: ignore[override]
        return {"type": type(self).__name__, "value": self.value}

    def __eq__(self, other: object) -> bool:
        return isinstance(other, type(self)) and other.value == self.value

    def __hash__(self) -> int:
        return id(self)

    def __repr__(self) -> str:
        return f"{type(self).__name__}(value={self.value})"

    def __str__(self) -> str:
        return f"{type(self).__name__}(value={self.value})"


class RuleWithListOfIntsOrRanges(RuleWithValue):
    def __init__(
        self,
        value: Optional[List[Union[int, Range]]] = None,
    ) -> None:
        # copy the incoming list
        super().__init__(list(value) if value is not None else [])


class RuleChar(RuleWithListOfIntsOrRanges):
    pass


class RuleCharExclude(RuleWithListOfIntsOrRanges):
    pass


class RuleEnd(Rule):
    pass


UnresolvedRule = Union[RuleChar, RuleCharExclude, RuleEnd, "Any"]

# RuleRefs should never be exposed to the end user.
ResolvedRule = Union[RuleCharExclude, RuleChar, RuleEnd]

@dataclass
class PrintOpts:
    pointers: Any
    colorize: Callable[[Union[str, int], str], str]
    show_position: bool = False

__all__ = [
    "PrintOpts",
    "Range",
    "ResolvedRule",
    "Rule",
    "RuleChar",
    "RuleCharExclude",
    "RuleEnd",
    "RuleWithListOfIntsOrRanges",
    "RuleWithValue",
    "UnresolvedRule",
    "ValidInput",
]

File: packages/gbnf-experiment/data/20260909T151143Z_170b737d/ported_implementation/src/grammar_graph/grammar_graph_types.ts
import { RuleRef } from "./rule_ref.ts";
import type { GraphPointer } from "./graph_pointer.ts";
import type { Pointers } from "./pointers.ts";

export { RuleRef };

export interface PrintOpts {
  pointers?: Pointers | Iterable<PrintablePointer>;
  colorize: (text: string | number, color: string) => string;
  show_position?: boolean;
}

/** The shape `print.ts` needs; `GraphNode` satisfies it, and so do test doubles. */
export interface PrintableNode {
  id: string;
  rule: UnresolvedRule;
  next: PrintableNode | null;
  print(opts: PrintOpts): string;
}

export interface PrintablePointer {
  node: PrintableNode;
  print(opts: PrintOpts): string;
}

export type Range = [number, number];

export class Rule {
  equals(other: unknown): boolean {
    return other instanceof (this.constructor as new (...args: never[]) => Rule);
  }

  toString(): string {
    return `${this.constructor.name}()`;
  }

  /** Mirrors the reference's `__dict__` property, used to build JSON keys. */
  toDict(): Record<string, unknown> {
    return {
      type: this.constructor.name,
    };
  }
}

export class RuleWithValue<T = unknown> extends Rule {
  value: T;

  constructor(value: T) {
    super();
    this.value = value;
  }

  override equals(other: unknown): boolean {
    return (
      other instanceof (this.constructor as new (...args: never[]) => RuleWithValue<T>) &&
      deep_equals(this.value, (other as RuleWithValue<T>).value)
    );
  }

  override toString(): string {
    return `${this.constructor.name}(value=${JSON.stringify(this.value)})`;
  }

  override toDict(): Record<string, unknown> {
    return {
      ...super.toDict(),
      value: this.value,
    };
  }
}

export class RuleWithListOfIntsOrRanges extends RuleWithValue<Array<number | Range>> {
  constructor(value: Array<number | Range> = []) {
    // The reference copies the incoming list in `__post_init__`; `build_rule_stack`
    // relies on being able to mutate `value` without touching the caller's array.
    super([...value]);
  }
}

export class RuleChar extends RuleWithListOfIntsOrRanges {}

export class RuleCharExclude extends RuleWithListOfIntsOrRanges {}

export class RuleEnd extends Rule {}

export type UnresolvedRule = RuleChar | RuleCharExclude | RuleRef | RuleEnd;

/**
 * ValidInput can either be a string, or a number indicating a code point.
 * It CANNOT be a number representing a number; a number intended as input (like "8")
 * should be passed in as a string.
 */
export type ValidInput = string | number | number[];

/** RuleRefs should never be exposed to the end user. */
export type ResolvedRule = RuleCharExclude | RuleChar | RuleEnd;

export type ResolvedGraphPointer = GraphPointer<ResolvedRule>;

const deep_equals = (a: unknown, b: unknown): boolean => {
  if (Array.isArray(a) && Array.isArray(b)) {
    return a.length === b.length && a.every((item, idx) => deep_equals(item, b[idx]));
  }
  return a === b;
};
